Issue #494: New workflow for creating and validating forecast objects

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #494.

This PR
- replaces the function `validate.default()` with a new class construction helper `as_forecast()`, a function that determines the forecast type, constructs a new class object, and validates that object. `as_forecast()` is a generic with a single method `as_forecast.default()` that does the work
- Replaces all instances of `validate()` (in the sense of the previous `validate.default()`) in the code, the news file, and the tests
- Reorganises the documentation for `score()` and `as_forecast()` a bit, such that a section that is common to both can be reused
- adds the .trackdown folder to .gitignore and .buildignore - the changes suggested in #494, but I recently used trackdown on a different branch (but also gitignored it so I can't just delete the file) for updates to the manuscript
- Renamed  the generic `validate()` to `validate_forecast()`. The class construction helper (which was previously `validate.default()` is therefore now `as_forecast()` and the validation function (previously the methods of `validate()`) are now methods of `validate_forecast()`

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [x] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

seabbs commented on NAMESPACE:
```diff
@@ -8,16 +9,16 @@ S3method(score,scoringutils_binary)
-S3method(validate,scoringutils_point)
-S3method(validate,scoringutils_quantile)
-S3method(validate,scoringutils_sample)
+S3method(validate_forecast,scoringutils_binary)
```
I thought we had renamed these to forecast_sample etc?

seabbs commented on R/validate.R:
```diff
@@ -45,13 +35,37 @@ validate.default <- function(data, ...) {
 #' @export
 #' @keywords check-forecasts
-validate.scoringutils_binary <- function(data, ...) {
+validate_forecast.scoringutils_binary <- function(data, ...) {
```
not forecast.binary?

seabbs commented on README.Rmd:
```diff
@@ -85,12 +85,13 @@ example_quantile %>%
 example_quantile %>%
   set_forecast_unit(c("location", "target_end_date", "target_type", "horizon", "model")) %>%
-  validate() %>%
+  as_forecast() %>%
```
Do we need/want this duplication of validation here?

nikosbosse replied:
Ah yeah, definitely not. Thanks.

seabbs commented on vignettes/scoringutils.Rmd:
````diff
@@ -61,21 +61,21 @@ set_forecast_unit(
 
 ```{r}
-validate(example_quantile)
+as_forecast(example_quantile)
````
to match the recommended workflow this should be assigned and then printed to check its contents. You can then use this new object (i.e of type forecast) in all of the following.

seabbs commented on vignettes/scoringutils.Rmd:
```diff
@@ -61,21 +61,21 @@ set_forecast_unit(
 If you are unsure what your input data should look like, have a look at the `example_quantile`, `example_integer`, `example_continuous` and `example_binary` data sets provided in the package. 
 
-The output of `validate()` can later be directly used as input to `score()` (otherwise, `score()` will just run `validate()` anyway internally). 
+The output of `as_forecast()` can later be directly used as input to `score()`. This is the recommended workflow (however, `score()` will run `as_forecast()` internally if this hasn't happened before). 
```
so why aren't you doing this below

nikosbosse commented on R/validate.R:
```diff
@@ -45,13 +35,37 @@ validate.default <- function(data, ...) {
 
   # validate class
-  validate(data)
+  validate_forecast(data)
```
You mean doing something like the following?
```
as_forecast.default <- function(data, validate = TRUE, ...) {
  assert(check_data_columns(data))

  # find forecast type
  forecast_type <- get_forecast_type(data)

  # construct class
  data <- new_scoringutils(data, paste0("scoringutils_", forecast_type))

  # validate class
  if (validate) {
    validate_forecast(data)
  }
}
```

and then 

```
score.default <- function(data, ...) {
  data <- as_forecast(data, validate = FALSE)
  score(data, ...)
}
```
I thought about something like that as well. I think the trade-off seems code complexity vs. a small computational overhead. We could also leave `as_forecast()` intact and just change `score.default()`: 

```
score.default <- function(data, ...) {
  # find forecast type
  forecast_type <- get_forecast_type(data)

  # construct class
  data <- new_scoringutils(data, paste0("scoringutils_", forecast_type))
  score(data, ...)
}

I could also live with the duplication just staying as it is. Also noting there is still #507.

seabbs replied:
Yes, that is what I meant. I think changing score.default makes the most sense but your right its not that much overhead.

nikosbosse replied:
okk. Done that!

seabbs commented:
Can you resolve and request another review when ready please1

nikosbosse commented:
I updated `score.default()` to not validate forecasts. I also updated the Vignette to make use `set_forecast_unit()` and `as_forecast()` consistently. 

We'll likely need another round of edits for the Vignette. I suggest to not mix that with this PR. I haven't finished updating the scoringutils manuscript yet and I think we can re-use parts of that, creating a coherent structure that makes sense. I opened https://github.com/epiforecasts/scoringutils/issues/532 to keep track of this. 

Side note: I also made a new comment to #435 since I noticed that the current default for `by` in `summarise_scores()` maybe doesn't that make much sense anymore now that we're not returning scores per quantile anymore. 

PR is good for another round of review. Thanks a lot!

seabbs reviewed (approved):
Nice. All makes sense. Looks good to me.
